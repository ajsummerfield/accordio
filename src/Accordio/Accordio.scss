.accordio {
  position: relative;
  $p: &;

  &__item {
    position: relative;
    overflow: hidden;

    &:not(:last-child) {
      margin-bottom: 15px;
    }

    &--active {

      #{$p}__panel {
        max-height: 1000px;
      }
    }
  }

  &__heading {
    position: relative;
    background-color: #cccccc;
    padding: 15px;
    cursor: pointer;
    transition: background-color linear 250ms;
    z-index: 15;

    &:hover {
      background-color: darken(#cccccc, 10%);
    }
  }

  &__panel {
    max-height: 0;
    transition: max-height ease-in-out 500ms;
  }

  &__content {
    padding: 15px;
    background-color: #eeeeee;
  }
}
