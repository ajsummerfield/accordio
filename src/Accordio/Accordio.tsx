import React, { useState } from 'react';
import AccordioContext from './context/AccordioContext';

type AccordioProps = {
  children: any;
};

const Accordio = ({ children }: AccordioProps) => {
  const [activeIndex, setActiveIndex] = useState(-1);
  const context = { activeIndex, setActiveIndex };

  return (
    <AccordioContext.Provider value={context}>
      <div className="accordio">{children}</div>
    </AccordioContext.Provider>
  );
};

export default Accordio;
