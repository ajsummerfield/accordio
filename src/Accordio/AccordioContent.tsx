import React, { useEffect, useRef, useState } from 'react';
import useAccordioContext from './hooks/useAccordioContext';
import useAccordioItemContext from './hooks/useAccordioItemContext';

type AccordioContentProps = {
  children: JSX.Element;
};

const AccordioContent = ({ children }: AccordioContentProps) => {
  const [height, setHeight] = useState(0);
  const { activeIndex } = useAccordioContext();
  const { index } = useAccordioItemContext();
  const panel = useRef(null);

  useEffect(() => {
    setHeight(activeIndex === index ? panel.current.scrollHeight : 0);
  }, [activeIndex]);

  return (
    <div
      className="accordio__panel"
      style={{ maxHeight: `${height}px` }}
      ref={panel}
    >
      <div className="accordio__content">{children}</div>
    </div>
  );
};

export default AccordioContent;
