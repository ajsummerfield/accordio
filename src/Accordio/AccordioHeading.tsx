import React from 'react';

type AccordioHeadingProps = {
  children: any;
};

const AccordioHeading = ({ children }: AccordioHeadingProps) => {
  return (
    <div className="accordio__heading">
      {children}
      <div className="accordio__chevron"></div>
    </div>
  );
};

export default AccordioHeading;
