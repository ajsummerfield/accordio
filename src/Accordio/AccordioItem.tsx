import React from 'react';
import AccordioItemContext from './context/AccordioItemContext';
import useAccordioContext from './hooks/useAccordioContext';
import useAccordioItemContext from './hooks/useAccordioItemContext';

type AccordioItemProps = {
  index: number;
  children: JSX.Element;
};

const AccordioItem = ({ index, children }: AccordioItemProps) => {
  const { activeIndex, setActiveIndex } = useAccordioContext();
  const { setIndex } = useAccordioItemContext();
  const className = `accordio__item 
    ${activeIndex === index ? 'accordio__item--active' : ''}`;

  setIndex(index);

  const handleOnClick = () => {
    setActiveIndex(activeIndex !== index ? index : null);
  };

  const context = {
    index: index,
    setIndex: setIndex,
  };

  return (
    <AccordioItemContext.Provider value={context}>
      <div className={className} onClick={handleOnClick}>
        {children}
      </div>
    </AccordioItemContext.Provider>
  );
};

export default AccordioItem;
