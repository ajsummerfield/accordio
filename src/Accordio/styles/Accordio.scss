.accordio {
  $p: &;
  position: relative;

  &__item {
    position: relative;
    overflow: hidden;

    &:not(:last-child) {
      margin-bottom: 15px;
    }

    &--active {

      #{$p}__heading {
        background: darken(#cccccc, 10%);
      }

      #{$p}__chevron {
        transform: translateY(-50%) rotate(0deg);
      }
    }
  }

  &__heading {
    position: relative;
    background-color: #cccccc;
    padding: 15px;
    cursor: pointer;
    transition: background-color linear 250ms;
    z-index: 15;

    &:hover {
      background-color: darken(#cccccc, 10%);
    }
  }

  &__chevron {
    position: absolute;
    top: 50%;
    right: 15px;
    transform: translateY(-50%) rotate(-90deg);
    height: 16px;
    width: 25px;
    border: 2px solid transparent;
    transition: all 250ms linear;

    &::after {
      content: '';
      display: block;
      position: absolute;
      top: -3px;
      left: 4px;
      transform: rotate(45deg);
      height: 10px;
      width: 10px;
      border-bottom: 2px solid #333333;
      border-right: 2px solid #333333;
    }
  }

  &__panel {
    max-height: 0;
    transition: max-height ease-in-out 500ms;
  }

  &__content {
    padding: 15px;
    background-color: #eeeeee;
  }
}
