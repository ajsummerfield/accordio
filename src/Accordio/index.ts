import Accordio from './Accordio';
import AccordioItem from './AccordioItem';
import AccordioHeading from './AccordioHeading';
import AccordioContent from './AccordioContent';

export {
  Accordio,
  AccordioItem,
  AccordioHeading,
  AccordioContent
};
