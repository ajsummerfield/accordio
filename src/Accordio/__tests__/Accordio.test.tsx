import React from 'react';
import Accordio from '../Accordio';
import { shallowRenderer } from './helpers';

describe('Accordio Component', () => {
  let component: JSX.Element;

  describe('when rendered', () => {
    beforeEach(() => {
      component = <Accordio>Hello World</Accordio>;
    });
  
    it('Should match the snapshot', () => {
      const tree = shallowRenderer(component);
      expect(tree).toMatchSnapshot();
    });
  });
});
