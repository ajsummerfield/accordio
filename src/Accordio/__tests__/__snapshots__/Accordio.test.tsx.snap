// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordio Component when rendered Should match the snapshot 1`] = `
<Context.Provider
  value={
    Object {
      "activeIndex": null,
      "setActiveIndex": [Function],
    }
  }
>
  <div
    className="accordio"
  >
    Hello World
  </div>
</Context.Provider>
`;
